mpc-core/src/garble/circuit.rs

```diff
@@ -1,82 +1,295 @@
+            encrypted_gates: self.encrypted_gates.clone(),
+            decoding: decoding.then(|| self.decoding()),
         }
     }
+
+    /// Validates that provided output labels are correct
+    pub fn validate_output(&self, output_labels: &[BinaryLabel]) -> Result<(), Error> {
+        if output_labels.len() != self.circ.output_count() {
+            return Err(Error::InvalidOutputLabels);
+        }
+        let pairs = self
+            .wire_labels
+            .iter()
+            .enumerate()
+            .skip(self.circ.len() - self.circ.output_len());
+
+        if output_labels.iter().zip(pairs).all(|(label, (id, pair))| {
+            (label.id == id)
+                & ((label.value == *pair[0].as_ref()) | (label.value == *pair[1].as_ref()))
+        }) {
+            Ok(())
+        } else {
+            Err(Error::InvalidOutputLabels)
+        }
+    }
+}
+
+/// Generates pairs of wire labels \[W_0, W_0 ^ delta\]
+pub fn generate_labels<R: Rng + CryptoRng>(
+    rng: &mut R,
+    delta: Option<&Block>,
+    count: usize,
+    offset: usize,
+) -> (Vec<[BinaryLabel; 2]>, Block) {
+    let delta = match delta {
+        Some(delta) => *delta,
+        None => {
+            let mut delta = Block::random(rng);
+            delta.set_lsb();
+            delta
+        }
+    };
+    let low = Block::random_vec(rng, count);
+    (
+        low.into_iter()
+            .enumerate()
+            .map(|(id, value)| {
+                [
+                    BinaryLabel {
+                        id: id + offset,
+                        value,
+                    },
+                    BinaryLabel {
+                        id: id + offset,
+                        value: value ^ delta,
+                    },
+                ]
+            })
+            .collect(),
+        delta,
+    )
+}
+
+/// Returns wire labels corresponding to wire truth values
+///
+/// Panics if wire id not in label collection
+pub fn choose_labels<T: Copy>(labels: &[[T; 2]], wires: &[usize], values: &[bool]) -> Vec<T> {
+    wires
+        .iter()
+        .zip(values.iter())
+        .map(|(id, value)| labels[*id][*value as usize])
+        .collect()
+}
+
+/// Clones an array of input labels, sorts them, and validates according to a circuit description.
+pub fn prepare_inputs(circ: &Circuit, inputs: &[BinaryLabel]) -> Result<Vec<BinaryLabel>, Error> {
+    let mut inputs = Vec::from(inputs);
+    if circ.input_len() != inputs.len() {
+        return Err(Error::InvalidInput(InputError::InvalidCount(
+            circ.input_len(),
+            inputs.len(),
+        )));
+    }
+
+    inputs.sort_by_key(|label| label.id);
+    inputs.dedup_by_key(|label| label.id);
+
+    if circ.input_len() != inputs.len() {
+        return Err(Error::InvalidInput(InputError::Duplicate));
+    }
+
+    Ok(inputs)
+}
+
+/// Decodes output wire labels into plaintext.
+///
+/// Using the point-and-permute (p&p) technique, the two adjacent labels
```

Comment:
I'm sorry, the wroding was confusing the first time around. Could we change this to:

// Thanks to the point-and-permute (p&p) technique, the two adjacent labels
/// will have the opposite p&p bits. We apply the decoding to the p&p bits.